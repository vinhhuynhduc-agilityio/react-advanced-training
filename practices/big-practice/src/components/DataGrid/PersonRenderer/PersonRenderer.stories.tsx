import type { Meta, StoryObj } from '@storybook/react';
import { PersonRenderer } from '.'; 
import { ICellRendererParams } from 'ag-grid-community';

// Mock Data
const mockUserData = {
  "id": "d290f1ee-6c54-4b01-90e6-d701748f0851",
  "fullName": "Joe Bloggs",
  "earnings": "$11500",
  "email": "john@example.com",
  "avatarUrl": "https://i.pravatar.cc/150?img=1",
  "registered": "May 21, 2020 17:02:06",
  "lastUpdated": "Nov 11, 2024 14:34:21"
};

// Metadata
const meta: Meta<typeof PersonRenderer> = {
  title: 'Components/PersonRenderer',
  component: PersonRenderer,
  tags: ['autodocs'],
  parameters: {
    layout: 'centered',
  },
};

export default meta;

type Story = StoryObj<typeof PersonRenderer>;

export const Default: Story = {
  args: {
    data: mockUserData,
  } as ICellRendererParams<typeof mockUserData>,
};
