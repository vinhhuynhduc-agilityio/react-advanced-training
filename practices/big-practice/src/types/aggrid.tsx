import { ColDef } from "ag-grid-community";
import { RowData } from "./users";

export interface DataGridProps {
  rowData: RowData[];
  columnDefs: ColDef<RowData>[];
};