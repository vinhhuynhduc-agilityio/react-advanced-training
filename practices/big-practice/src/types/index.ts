export * from './chart';
export * from './task';
export * from './user';
export * from './project';
