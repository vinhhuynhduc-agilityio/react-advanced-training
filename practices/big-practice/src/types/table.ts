export interface UserData {
  id: string;
  fullName: string;
  earnings: string;
  email: string;
  avatarUrl: string;
  registered: string;
  lastUpdated: string;
}

export interface TaskData {
  id: string;
  userId: string;
  projectId: string;
  taskName: string;
  startDate: string;
  completedDate: string;
  currency: number;
  status: boolean;
  projectName: string;
  fullName: string;
  tooltipShowDelay?: number;
}

export interface ProjectsData {
  id: string;
  projectName: string;
}

export interface UserListDrawerProps {
  users: UserData[];
  selectedUserId: string | null;
  onUserSelected: (userId: string | null) => void;
  sourceComponent: string | null;
}

export type StatusWithDate = {
  status: boolean;
  completedDate: string;
};

export type FieldValue =
  | string
  | number
  | boolean
  | ProjectsData
  | UserData
  | StatusWithDate;
