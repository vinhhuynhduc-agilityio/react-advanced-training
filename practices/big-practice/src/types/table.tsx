import {
  ColDef,
  GetRowIdParams,
  GridReadyEvent,
  RowClassParams,
  RowClickedEvent
} from "ag-grid-community";

export interface UserData {
  id: string;
  fullName: string;
  earnings: string;
  email: string;
  avatarUrl: string;
  registered: string;
  lastUpdated: string;
};

export interface TaskData {
  id: string;
  userId: string;
  projectId: string;
  taskName: string;
  startDate: string;
  completedDate: string;
  currency: number;
  status: boolean;
  projectName: string;
  fullName: string;
};

export interface ProjectsData {
  id: string;
  projectName: string;
};

export interface DataGridProps<T> {
  rowData: T[];
  userData?: T[];
  columnDefs: ColDef<T>[];
  onRowClicked?: (event: RowClickedEvent) => void;
  getRowClass?: (params: RowClassParams) => string;
  rowHeight?: number;
  onGridReady: (event: GridReadyEvent) => void;
  getRowId: (params: GetRowIdParams<T>) => string;
};

export interface UserListDrawerProps {
  users: UserData[];
  selectedUserId: string | null;
  onUserSelected: (userId: string | null) => void;
  sourceComponent: string | null;
};

export interface TaskDataProps {
  tasks: TaskData[];
  selectedUserId: string | null;
  onTaskRowSelected: (userId: string | null) => void;
  projects: ProjectsData[];
  sourceComponent: string | null;
  users: UserData[]
};
