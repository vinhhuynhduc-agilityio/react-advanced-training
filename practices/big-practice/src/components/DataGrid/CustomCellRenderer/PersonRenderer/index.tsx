import { UserData } from "@/types";
import { ICellRendererParams } from "ag-grid-community";

export const PersonRenderer = (params: ICellRendererParams<UserData>) => {
  if (!params.data) {
    return null;
  }

  const { avatarUrl, fullName, earnings } = params.data;
  return (
    <div className="flex items-center py-2">
      <img
        src={avatarUrl}
        alt="avatar"
        className="w-14 h-14 rounded-full mr-4"
      />
      <div className="flex flex-col">
        <div className="text-base font-medium text-[#313131]">{fullName}</div>
        <div className="text-base font-normal text-[#475466]">{earnings}</div>
      </div>
    </div>
  );
};
