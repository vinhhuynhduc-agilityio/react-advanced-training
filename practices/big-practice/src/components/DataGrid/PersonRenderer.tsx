import { ICellRendererParams } from "ag-grid-community";
import { RowData } from "../../types/users";

export const PersonRenderer = (params: ICellRendererParams<RowData>) => {
  if (!params.data) {
    return null;
  }

  const { avatar_url, full_name, earnings } = params.data;
  return (
    <div className="flex items-center py-2">
      <img
        src={avatar_url}
        alt="avatar"
        className="w-14 h-14 rounded-full mr-4"
      />
      <div className="flex flex-col">
        <div className="text-base font-medium">{full_name}</div>
        <div className="text-base font-normal">{earnings}</div>
      </div>
    </div>
  );
};
