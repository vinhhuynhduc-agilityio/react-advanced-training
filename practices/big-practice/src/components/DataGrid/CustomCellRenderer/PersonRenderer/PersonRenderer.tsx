import { UserData } from "@/types/table";
import { ICellRendererParams } from "ag-grid-community";

export const PersonRenderer = (params: ICellRendererParams<UserData>) => {
  if (!params.data) {
    return null;
  }

  const { avatarUrl, fullName, earnings } = params.data;
  return (
    <div className="flex items-center py-2">
      <img
        src={avatarUrl}
        alt="avatar"
        className="w-14 h-14 rounded-full mr-4"
      />
      <div className="flex flex-col">
        <div className="text-base font-medium">{fullName}</div>
        <div className="text-base font-normal">{earnings}</div>
      </div>
    </div>
  );
};
