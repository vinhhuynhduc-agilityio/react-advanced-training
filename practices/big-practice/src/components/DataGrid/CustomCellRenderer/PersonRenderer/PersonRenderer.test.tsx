import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { ICellRendererParams } from 'ag-grid-community';
import { UserData } from '@/types/table';
import { PersonRenderer } from './PersonRenderer';

const mockData: UserData = {
  "id": "d290f1ee-6c54-4b01-90e6-d701748f0851",
  "fullName": "Joe Bloggs",
  "earnings": "$5000",
  "email": "john@example.com",
  "avatarUrl": "https://i.pravatar.cc/150?img=1",
  "registered": "May 21, 2020 17:02:06",
  "lastUpdated": "October 10, 2023 12:30:00"
};

describe('PersonRenderer component', () => {
  const params: ICellRendererParams<UserData> = { data: mockData } as ICellRendererParams<UserData>;
  it('matches snapshot for default state', () => {
    const { container } = render(<PersonRenderer {...params} />);
    expect(container).toMatchSnapshot();
  });

  it('should render user data correctly', () => {
    const params: ICellRendererParams<UserData> = { data: mockData } as ICellRendererParams<UserData>;

    render(<PersonRenderer {...params} />);

    expect(screen.getByText('Joe Bloggs')).toBeInTheDocument();
    expect(screen.getByText('$5000')).toBeInTheDocument();
    const img = screen.getByAltText('avatar') as HTMLImageElement;
    expect(img.src).toBe('https://i.pravatar.cc/150?img=1');
  });
});
