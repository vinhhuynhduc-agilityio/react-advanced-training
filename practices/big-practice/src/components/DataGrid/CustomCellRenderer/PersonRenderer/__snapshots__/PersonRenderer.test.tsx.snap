// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonRenderer component matches snapshot for default state 1`] = `
<div>
  <div
    class="flex items-center py-2"
  >
    <img
      alt="avatar"
      class="w-14 h-14 rounded-full mr-4"
      src="https://i.pravatar.cc/150?img=1"
    />
    <div
      class="flex flex-col"
    >
      <div
        class="text-base font-medium text-[#313131]"
      >
        Joe Bloggs
      </div>
      <div
        class="text-base font-normal text-[#475466]"
      >
        $5000
      </div>
    </div>
  </div>
</div>
`;
